### PR TITLE
Add .editorconfig

### PR DESCRIPTION
**Pull request by iSazonov**
## PR Summary

Related #4708.

The .editorconfig file can be used by various editors and tools like Visual Studio and CodeFormatter.

The .editorconfig comes from Roslyn and CoreFX repositories:
https://github.com/dotnet/corefx/blob/master/.editorconfig
https://github.com/dotnet/roslyn/blob/master/.editorconfig

Description http://EditorConfig.org

Options https://docs.microsoft.com/en-us/visualstudio/ide/editorconfig-code-style-settings-reference

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [x] Not Applicable or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+dotnet_naming_symbols.static_fields.applicable_kinds   = field
+dotnet_naming_symbols.static_fields.required_modifiers = static
+
+dotnet_naming_style.static_prefix_style.required_prefix = s_
+dotnet_naming_style.static_prefix_style.capitalization = camel_case
+
+# Internal and private fields should be _camelCase
+dotnet_naming_rule.camel_case_for_private_internal_fields.severity = suggestion
+dotnet_naming_rule.camel_case_for_private_internal_fields.symbols  = private_internal_fields
+dotnet_naming_rule.camel_case_for_private_internal_fields.style    = camel_case_underscore_style
```
Comment (TravisEz13) [inline .editorconfig]: This isn't in our style: https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/coding-guidelines.md

**Comment by TravisEz13** [comment]
some of these conflict with style cop rules.  It would be good to make them consistent (disable the conflicting rules)

File: .editorconfig (hunk 1 above)
Comment (iSazonov) [reply .editorconfig]: Could you please clarify? We have in the guideline:
> Use _camelCase to name internal and private fields and use readonly where possible.

**Comment by iSazonov** [comment]
I believe we have already gained experience with StyleCop and we can turn off a few more rules. But first I would like to make some PR with CodeFormatting and style fixes so that we see the state of things more clearly. Also I guess CodeFormatter will use the `.editorconfig` file.

File: .editorconfig (hunk 1 above)
Comment (TravisEz13) [reply .editorconfig]: I must have misread the guidance.  Resolved

**Comment by TravisEz13** [comment]
Many of the naming rules, especially for non-public fields, conflict with the style cop rules.  I believe those rules should be disabled before we add a file that defines conflicting rules.

**Comment by iSazonov** [comment]
@SteveL-MSFT Could you please share plans about fixing style issues?
My suggetion is to spend 1-2 weeks and fix as many style issues as we need, adjust and approve StyleCop and .editorconfig rules, reformat all code with CodeFormatter. After that we can return to normal contributing process with minimal style issue pressure.

**Comment by SteveL-MSFT** [comment]
@iSazonov I think after GA of 6.1 would be the best time to bulk address style issues to have consistency.  We're still targeting mid Sept for GA, so perhaps we can postpone this until then.

**Comment by anmenaga** [comment]
@SteveL-MSFT GA shipped; it's time to look at this. :)

**Comment by iSazonov** [comment]
Rebased to update CIs.

File: .editorconfig [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+# Top-most EditorConfig file
+root = true
+
+# Don't use tabs for indentation.
+[*]
+indent_style = space
+# (Please don't specify an indent_size here; that has too many unintended consequences.)
+
+# Code files
+[*.{cs,csx,vb,vbx}]
```
Comment (SteveL-MSFT) [inline .editorconfig]: We don't have any vb code, but we do have cpp.

File: .editorconfig [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+indent_size = 4
+insert_final_newline = true
+charset = utf-8
+
+# Xml project files
+[*.{csproj,vbproj,vcxproj,vcxproj.filters,proj,projitems,shproj}]
```
Comment (SteveL-MSFT) [inline .editorconfig]: what about ps1xml files?  Should we remove extensions we don't use in this repo?

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

File: .editorconfig (hunk 2 above)
Comment (iSazonov) [reply .editorconfig]: Fixed.

File: .editorconfig (hunk 3 above)
Comment (iSazonov) [reply .editorconfig]: Added resx,ps1xml. Removed unused.

**Review by SteveL-MSFT — APPROVED** [review]

**Comment by iSazonov** [comment]
@TravisEz13 Can we merge? Or we should address
> some of these conflict with style cop rules. It would be good to make them consistent (disable the conflicting rules)

If .editorconfig rules is ok we could continue with style cop rules in follow PRs.

**Pull request merged**